carma_ros2_utils/src/lifecycle_component_wrapper.cpp

```diff
@@ -0,0 +1,410 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * Modifications copyright (C) 2021 Leidos
+ * - Converted into Lifecycle Component Wrapper
+ * 
+ */ 
+
+#include "carma_ros2_utils/lifecycle_component_wrapper.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <ament_index_cpp/get_resource.hpp>
+#include <class_loader/class_loader.hpp>
+#include <rcpputils/filesystem_helper.hpp>
+#include <rcpputils/split.hpp>
+
+using namespace std::placeholders;
+
+namespace carma_ros2_utils
+{
+
+///// CARMA CHANGE /////
+LifecycleComponentWrapper::LifecycleComponentWrapper(const rclcpp::NodeOptions & node_options)
+: carma_ros2_utils::CarmaLifecycleNode(node_options)
+{
+  // Do nothing
+}
+
+void LifecycleComponentWrapper::initialize(std::weak_ptr<rclcpp::Executor> executor) {
+  executor_ = executor;
+
+  loadNode_srv_ = create_service<LoadNode>(
+    "~/_container/load_node",
+    std::bind(&LifecycleComponentWrapper::on_load_node, this, _1, _2, _3, boost::none));
+
+  unloadNode_srv_ = create_service<UnloadNode>(
+    "~/_container/unload_node",
+    std::bind(&LifecycleComponentWrapper::on_unload_node, this, _1, _2, _3));
+
+  listNodes_srv_ = create_service<ListNodes>(
+    "~/_container/list_nodes",
+    std::bind(&LifecycleComponentWrapper::on_list_nodes, this, _1, _2, _3));
+
+}
+
+////// END CARMA CHANGE /////
+
+LifecycleComponentWrapper::~LifecycleComponentWrapper()
+{
+  if (node_wrappers_.size()) {
+    RCLCPP_DEBUG(get_logger(), "Removing components from executor");
+    if (auto exec = executor_.lock()) {
+      for (auto & wrapper : node_wrappers_) {
+        exec->remove_node(wrapper.second.get_node_base_interface());
+      }
+    }
+  }
+}
+
+std::vector<LifecycleComponentWrapper::ComponentResource>
+LifecycleComponentWrapper::get_component_resources(
+  const std::string & package_name, const std::string & resource_index) const
+{
+  std::string content;
+  std::string base_path;
+  if (
+    !ament_index_cpp::get_resource(
```

Comment:
Is this a ROS2 cpp library for accessing files based on the installed location of the ROS package? Has something like that existed in ROS1 or is that new? Neat.

Reply:
I believe most of this is new to ROS2 and its made possible via the component registration macros you might have seen in the system controller PRs